WordPress/Classes/Utility/App Configuration/WidgetConfiguration.swift

```diff
@@ -0,0 +1,27 @@
+// WordPress Widget configuration
+
+import Foundation
+
+@objc extension AppConfiguration {
+
+    @objc(AppConfigurationWidget)
+    class Widget: NSObject {
+        @objc static let statsTodayWidgetKeychainTokenKey = "OAuth2Token"
+        @objc static let statsTodayWidgetKeychainServiceName = "TodayWidget"
+        @objc static let statsTodayWidgetUserDefaultsSiteIdKey = "WordPressTodayWidgetSiteId"
+        @objc static let statsHomeWidgetsUserDefaultsSiteIdKey = "WordPressHomeWidgetsSiteId"
+        @objc static let statsHomeWidgetsUserDefaultsLoggedInKey = "WordPressHomeWidgetsLoggedIn"
+        @objc static let statsTodayWidgetUserDefaultsSiteNameKey = "WordPressTodayWidgetSiteName"
+        @objc static let statsTodayWidgetUserDefaultsSiteUrlKey = "WordPressTodayWidgetSiteUrl"
+        @objc static let statsTodayWidgetUserDefaultsSiteTimeZoneKey = "WordPressTodayWidgetTimeZone"
+        @objc static let homeWidgetTodayKind = "WordPressHomeWidgetToday"
+        @objc static let homeWidgetAllTimeKind = "WordPressHomeWidgetAllTime"
+        @objc static let homeWidgetThisWeekKind = "WordPressHomeWidgetThisWeek"
+        @objc static let homeWidgetTodayProperties = "WordPressHomeWidgetTodayProperties"
+        @objc static let homeWidgetAllTimeProperties = "WordPressHomeWidgetAllTimeProperties"
+        @objc static let homeWidgetThisWeekProperties = "WordPressHomeWidgetThisWeekProperties"
+        @objc static let homeWidgetTodayFilename = "HomeWidgetTodayData.plist"
+        @objc static let homeWidgetAllTimeFilename = "HomeWidgetAllTimeData.plist"
+        @objc static let homeWidgetThisWeekFilename = "HomeWidgetThisWeekData.plist"
+    }
+}
```

Comment:
Thank you for your comments.

> Have you thought about how this approach of duplicating a class definition will scale?

Yes, I thought about it and this is not ideal. Your example is excellent. I had experience in other larger projects where we had to support multiple targets with different configurations. The use case was a bit different than we have with WordPress and Jetpack but technically the problem looks similar.

- The worst approach was having multiple control statements scattered in the codebase. This made understanding the differences between different configurations hard and any changes needed to be made directly in particular classes.
- We did try duplicating class definitions but we soon hit the problem you were describing. It's also not 100% transparent as well, which files are configuration files and which ones are included in which targets.
- It's possible to move the configuration to `xcconfig` configuration files. It would allow us to have all the configuration, including build settings, in one place. Although that wouldn't guarantee we set all the required fields for all configurations. It could be used in tandem with some approaches in the code.
- Our configurations could conform to one `protocol` and the correct configuration be resolved at runtime. This way both configurations could be included in both targets, allowing us to show build error when one of the configurations doesn't conform to a `protocol`.  We ended up using a flavor of this approach in our previous project.

All in all, I think it deserves a wider discussion, how we could set up a configuration in a way that would support our needs going forward. Otherwise, we could end up with a messy solution.

Reply:
> All in all, I think it deserves a wider discussion, how we could set up a configuration in a way that would support our needs going forward. Otherwise, we could end up with a messy solution.

Definitely. I'd be great to address the problem before it happens, but the kind of changes we are discussing are too far-reaching and it would make this PR go stale.

I'm glad to hear about the `protocol` idea, because that's something I'd been thinking about as well. We'd then need something that at runtime allocates the correct one but that seems simpler to achieve than keeping duplicated classes in sync.

Alternatively, and depending on what the plan long term is for these widgets, it might be beneficial to extract the bulk of the logic in a framework, then have each widget target be a thin wrapper to it and pass a `struct` with configurations. That would centralize things big time, but does come with an extra layer of complexity in isolating stuff.

We'll see...